### PR TITLE
ostest: Implement mutex move test functionality

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary

This change enhances NuttX's test coverage by validating mutex behavior
in non-standard relocation scenarios, while explicitly noting this
extends beyond POSIX specifications. The test helps prevent regressions

    in edge-case mutex handling.
    * Add test case for mutex move functionality in ostest
    * Verifies correct behavior when a mutex is relocated in memory
    * Introduces new test structure with thread argument passing
    * Includes validation of mutex state across thread operations
    * Related to Rust language discussion: https://github.com/rust-lang/rust/pull/138400
## Impact
    * New feature added?: YES - Adds new mutex move test scenario
    * Impact on user?: NO - Test suite enhancement only
    * Impact on build?: NO - Existing test infrastructure used
    * Impact on hardware?: NO - Pure software validation
    * Impact on documentation?: NO - No user-facing docs changed
    * Impact on security?: NO - Test-only code
    * Impact on compatibility?: NO - Maintains existing POSIX compliance
## Testing
ostest:
```
user_main: mutex test
Initializing mutex
Starting thread 1
Starting thread 2
                Thread1 Thread2
        Loops   32      32
        Errors  0       0

Testing moved mutex
Starting moved mutex thread 1
Starting moved mutex thread 2
                Thread1 Thread2
        Moved Loops     32      32
        Moved Errors    0       0

End of test memory usage:
VARIABLE  BEFORE   AFTER
======== ======== ========
arena     1fc5ee0  1fc5ee0
ordblks         4        2
mxordblk  1fbacc0  1fc07e8
uordblks     ad30     5678
fordblks  1fbb1b0  1fc0868

```

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



Yes, this PR appears to meet the NuttX requirements.  The summary clearly explains the change, its purpose, and how it works. The impact section adequately addresses all points, and the testing section provides evidence of validation, including before and after logs (although minimal). The reference to the Rust discussion is helpful context but not a direct NuttX issue/PR.

File: testing/ostest/mutex.c [hunk 1]
```diff
@@ -103,11 +116,172 @@ static void *thread_func(FAR void *parameter)
...
+  thread2_args.mutex_ptr = &moved_mutex;
+  thread2_args.mutex_value = &moved_mutex_value;
+  thread2_args.nloops = moved_nloops;
+  thread2_args.nerrors = moved_nerrors;
+  thread2_args.thread_id = 2;
+
+  /* Start two threads using the moved mutex */
+
+  printf("Starting moved mutex thread 1\n");
+#ifdef SDCC
```
Comment (xiaoxiang781216) [inline testing/ostest/mutex.c:241]: why do SDCC specially

File: testing/ostest/mutex.c [hunk 2]
```diff
@@ -103,11 +116,172 @@ static void *thread_func(FAR void *parameter)
...
+                         &thread1_args);
+#endif
+  if (status != 0)
+    {
+      printf("ERROR in moved mutex thread#1 creation\n");
+      ASSERT(false);
+    }
+
+  printf("Starting moved mutex thread 2\n");
+#ifdef SDCC
```
Comment (xiaoxiang781216) [inline testing/ostest/mutex.c:256]: ditto

File: testing/ostest/mutex.c [hunk 3]
```diff
@@ -103,11 +116,172 @@ static void *thread_func(FAR void *parameter)
...
+  status = pthread_create(&thread2, NULL, moved_mutex_thread_func,
+                         &thread2_args);
+#endif
+  if (status != 0)
+    {
+      printf("ERROR in moved mutex thread#2 creation\n");
+      ASSERT(false);
+    }
+
+#ifdef SDCC
```
Comment (xiaoxiang781216) [inline testing/ostest/mutex.c:269]: ditto

File: testing/ostest/mutex.c [hunk 4]
```diff
@@ -103,11 +116,172 @@ static void *thread_func(FAR void *parameter)
...
 }
 
+static void *moved_mutex_thread_func(FAR void *parameter)
+{
+  FAR mutex_thread_args_t *args = (FAR mutex_thread_args_t *)parameter;
+
+  int id = args->thread_id;
+  int ndx = id - 1;
+  int i;
+  unsigned long local_loops = 0;
```
Comment (xiaoxiang781216) [inline testing/ostest/mutex.c:126]:
```suggestion
  i local_loops;
```

Comment (hartmannathan) [reply testing/ostest/mutex.c:126]: Something doesn't seem right with this suggestion.

Did you mean:

```suggestion
  int local_loops;
```

?

File: testing/ostest/mutex.c [hunk 5]
```diff
@@ -103,11 +116,172 @@ static void *thread_func(FAR void *parameter)
...
- * Public Functions
+ * Name: mutex_move_test
+ *
+ * Description:
+ *   Test the mutex move functionality.  This test creates a mutex, moves
+ *   it to a new location, and then starts two threads that use the moved
+ *   mutex.
+ *   POSIX specification does not define the behavior of a mutex that is
+ *   moved.  This test is intended to verify that the mutex can be moved,
+ *   it's useful for some cases, see the discussion in
```
Comment (hartmannathan) [inline testing/ostest/mutex.c:179]:
```suggestion
 *   which is useful for some cases, see the discussion in
```

File: testing/ostest/mutex.c (hunk 4 above)
Comment (hartmannathan) [reply testing/ostest/mutex.c:126]: By the way, if changing `local_loops` to `int` here, then `nloops` of `mutex_thread_args_t` needs to be `int` also? Or is it better to leave as unsigned long?

File: testing/ostest/mutex.c [hunk 6]
```diff
@@ -103,11 +116,172 @@ static void *thread_func(FAR void *parameter)
...
+          printf("ERROR moved mutex thread=%d: "
+                 "moved_mutex_value should be zero, "
+                 "instead moved_mutex_value=%d\n",
+                 id, *(args->mutex_value));
+          ASSERT(false);
+          args->nerrors[ndx]++;
+        }
+
+      *(args->mutex_value) = 1;
+      for (i = 0; i < 10; i++)
```
Comment (hartmannathan) [inline testing/ostest/mutex.c:149]: Above, in the outer loop, we are using NLOOPS define; should there be a similar define for this loop, e.g., `NYIELDS`?

File: testing/ostest/mutex.c (hunk 4 above)
Comment (no1wudi) [reply testing/ostest/mutex.c]: All types changed to "int", I guess it's OK for such a simple test

Comment (xiaoxiang781216) [reply testing/ostest/mutex.c]: > Something doesn't seem right with this suggestion.
> 
> Did you mean:
> 
> ?

Yes, the reason is that NLOOPS is type of int const.

**Review by acassis — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]
